kratos/includes/kratos_parameters.h

```diff
@@ -54,969 +51,1541 @@ namespace Kratos {
+    public:
+        ///@name Life Cycle
+        ///@{
+
+        /**
+         * @brief Default constructor (iterator + root Parameter)
+         * @param itValue The iterator to adapt
+         * @param pRoot The root Parameter pointer
+         */
+        iterator_adaptor(value_iterator itValue, nlohmann::json* pValue,  Kratos::shared_ptr<nlohmann::json> pRoot) :mValueIterator(itValue), mrValue(*pValue), mpParameters(new Parameters(itValue, pValue, pRoot)) {}
+
+        /**
+         * @brief Default constructor (just iterator)
+         * @param itValue The iterator to adapt
+         */
+        iterator_adaptor(const iterator_adaptor& itValue) : mValueIterator(itValue.mValueIterator), mrValue(itValue.mrValue),  mpParameters(new Parameters(itValue->GetUnderlyingStorage(), itValue->GetUnderlyingRootStorage())) {}
+
+        ///@}
+        ///@name Operators
+        ///@{
+
+        /**
+         * @brief operator ++
+         * @details This adds one to the current iterator
+         * @return The next iterator
+         */
+        iterator_adaptor& operator++()
+        {
+            mValueIterator++;
+            return *this;
+        }
+
+        /**
+         * @brief operator ++int
+         * @details This adds N to the current iterator
+         * @param int N increment of iterations
+         * @return The +N iterator
+         */
+        iterator_adaptor operator++(int)
+        {
+            iterator_adaptor tmp(*this);
+            operator++();
+            return tmp;
+        }
+
+        /**
+         * @brief operator ==
+         * @details This operator check if the iterator is equal to another given iterator
+         * @return True if equal, false otherwise
+         */
+        bool operator==(const iterator_adaptor& rhs) const
+        {
+            return mValueIterator == rhs.mValueIterator;
+        }
+
+        /**
+         * @brief operator !=
+         * @details This operator check if the iterator is not equal to another given iterator
+         * @return True if not equal, false otherwise
+         */
+        bool operator!=(const iterator_adaptor& rhs) const
+        {
+            return mValueIterator != rhs.mValueIterator;
+        }
+
+        /**
+         * @brief operator*
+         * @details This operator returns the pointer of a given iterator
+         * @return The Pointer of the given iterator
+         */
+        Parameters& operator*() const
+        {
+            if (mValueIterator != mrValue.end())
+                mpParameters->mpValue = &(*mValueIterator);
+            return *mpParameters;
+        }
+
+        /**
+         * @brief operator ->
+         * @details This operator acces to the pointer of the Parameter
+         * @return The pointer of the parameter
+         */
+        Parameters* operator->() const
+        {
+            if (mValueIterator != mrValue.end())
+                mpParameters->mpValue = &(*mValueIterator);
+            return mpParameters.get();
```

Comment:
No, because when you do a ++ iteration your iterator becames end().Then you don't do anything but it is completely normal to reach the end() when doing a ++ operation

(Try to put an error and you will see)